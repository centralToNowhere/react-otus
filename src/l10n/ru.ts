export const l10n = {
  maxWidthLabel: "Макс. ширина:",
  maxHeightLabel: "Макс. высота:",
  cellSizeLabel: "Размер ячейки:",
  capacityLabel: "Процент заполненности:",
  speedLabel: "Обновлений в секунду:",
  buttonStart: "Старт",
  buttonStop: "Стоп",
  buttonReset: "Сброс",
  buttonGenerateRandom: "Создать",

  // Экран регистрации
  gameNameFull: "Conway's Game of Life",
  gameHeadingPart1: "Conway's",
  gameHeadingPart2: "Game of Life",
  buttonStartGameAsPlayer: "Начать игру",
  rulesHeading: "Правила",
  description1: "Игра «Жизнь» (англ. ",
  source: "Источник:",
  sourceName: "Игра «Жизнь»",
  description2:
    ") — клеточный автомат, придуманный английским математиком Джоном Конвеем в 1970 году.",
  rule1:
    "Место действия этой игры — «вселенная» — это размеченная на клетки поверхность или плоскость — безграничная, ограниченная, или замкнутая (в пределе — бесконечная плоскость).",
  rule2:
    "Каждая клетка на этой поверхности может находиться в двух состояниях: быть «живой» (заполненной) или быть «мёртвой» (пустой). Клетка имеет восемь соседей, окружающих её.",
  rule3:
    "Распределение живых клеток в начале игры называется первым поколением. Каждое следующее поколение рассчитывается на основе предыдущего по таким правилам:",
  rule3_1:
    "в пустой (мёртвой) клетке, рядом с которой ровно три живые клетки, зарождается жизнь",
  rule3_2:
    "если у живой клетки есть две или три живые соседки, то эта клетка продолжает жить; в противном случае, если соседей меньше двух или больше трёх, клетка умирает («от одиночества» или «от перенаселённости»)",
  rule4: "Игра прекращается, если:",
  rule4_1: "на поле не останется ни одной «живой» клетки",
  rule4_2:
    "конфигурация на очередном шаге в точности (без сдвигов и поворотов) повторит себя же на одном из более ранних шагов (складывается периодическая конфигурация)",
  rule4_3:
    "при очередном шаге ни одна из клеток не меняет своего состояния (складывается стабильная конфигурация; предыдущее правило, вырожденное до одного шага назад)",

  // Контейнер игрока
  registerPlayerLabel: "Введите имя игрока",
  logoutButton: "Выйти",
  generation: "Поколение",

  // Ошибки
  minCellsAmount: "Размер игрового поля меньше размера ячейки.",
  maxCellsAmount: "Превышено максимальное количество ячеек",
  positiveNumber: "Ожидается положительное число.",
  nonNegativeNumber: "Ожидается неотрицательное число.",
};
