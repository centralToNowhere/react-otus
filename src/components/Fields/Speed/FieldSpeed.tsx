import React from "react";
import { l10n } from "@/l10n/ru";
import { InputPatterns } from "@/components/Fields";
import { InputField, LabelField } from "@/components/Fields";
import { FormField } from "@/components/Form/FormField";
import { FieldError } from "@/components/Fields/FieldError/FieldError";
import { IFieldProps } from "@/components/Form";

export const FieldSpeed: React.FC<IFieldProps<"speed">> = (
  props: IFieldProps<"speed">
) => {
  return (
    <>
      <FormField>
        <LabelField htmlFor="speed-change">{l10n.speedLabel}</LabelField>
        <InputField
          id="speed-change"
          type="number"
          pattern={InputPatterns.float}
          step="0.1"
          name="speedChange"
          value={props.value}
          autoComplete="off"
          onChange={props.onChange}
          onBlur={props.onBlur}
        />
        <FieldError show={props.error.show} msg={props.error.msg} />
      </FormField>
    </>
  );
};
