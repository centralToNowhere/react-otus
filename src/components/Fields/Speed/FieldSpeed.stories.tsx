import React from "react";
import { Story, Meta } from "@storybook/react";
import { FieldSpeed } from "@/components/Fields/Speed/FieldSpeed";
import { IFieldProps } from "@/components/Form";
import { isValidPositiveNumericString } from "@/utils";

export default {
  title: "Components/FormFields/Speed",
  component: FieldSpeed,
  args: {
    formRawData: {
      rawSpeed: "2",
    },
    onRawChange: () => {
      // empty function
    },
    onChange: () => {
      // empty function
    },
    formValidators: [isValidPositiveNumericString()],
    error: {
      show: false,
      msg: "speed error",
    },
  },
} as Meta;

const Template: Story<
  IFieldProps<{
    rawSpeed: string;
  }>
> = (args) => {
  return <FieldSpeed {...args} />;
};

export const SpeedInput = Template.bind({});
