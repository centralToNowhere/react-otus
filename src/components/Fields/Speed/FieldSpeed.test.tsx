import React from "react";
import { render, screen } from "@testing-library/react";
import { FieldSpeed } from "@/components/Fields/Speed/FieldSpeed";
import { l10n } from "@/l10n/ru";

describe("FieldSpeed tests", () => {
  it("should render speed input", () => {
    const { asFragment } = render(
      <FieldSpeed
        value={"2"}
        onChange={() => {
          // empty function
        }}
        onBlur={() => {
          // empty function
        }}
        error={{
          show: false,
          msg: "Expected positive number",
        }}
      />
    );

    const input: HTMLInputElement = screen.getByLabelText(l10n.speedLabel);

    expect(input).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
