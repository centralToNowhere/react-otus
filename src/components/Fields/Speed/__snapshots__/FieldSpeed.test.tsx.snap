// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldSpeed tests should render speed input 1`] = `
<DocumentFragment>
  <div
    class="css-mkf0o6"
  >
    <label
      class="css-1wnvjz2"
      for="speed-change"
    >
      Обновлений в секунду:
    </label>
    <input
      autocomplete="off"
      class="css-16tc96n"
      id="speed-change"
      name="speedChange"
      pattern="([0-9]*[.])?[0-9]+"
      step="0.1"
      type="number"
      value="2"
    />
  </div>
</DocumentFragment>
`;
